addOns/kotlin/src/main/java/org/zaproxy/zap/extension/kotlin/KotlinEngineWrapper.java

```diff
@@ -0,0 +1,128 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.zap.extension.kotlin;
+
+import java.util.Collections;
+import java.util.List;
+import javax.script.ScriptContext;
+import javax.script.ScriptEngine;
+import javax.script.ScriptException;
+import javax.swing.ImageIcon;
+import kotlin.script.experimental.jsr223.KotlinJsr223DefaultScriptEngineFactory;
+import kotlin.script.experimental.jsr223.KotlinJsr223DefaultScriptEngineFactoryKt;
+import org.fife.ui.rsyntaxtextarea.SyntaxConstants;
+import org.zaproxy.zap.control.ExtensionFactory;
+import org.zaproxy.zap.extension.script.DefaultEngineWrapper;
+
+public class KotlinEngineWrapper extends DefaultEngineWrapper {
+
+    static {
+        System.setProperty(
+                KotlinJsr223DefaultScriptEngineFactoryKt
+                        .KOTLIN_JSR223_RESOLVE_FROM_CLASSLOADER_PROPERTY,
+                "true");
+    }
+
+    private ClassLoader classLoader;
+
+    public KotlinEngineWrapper() {
+        this(null);
+    }
+
+    public KotlinEngineWrapper(ClassLoader engineClassLoader) {
+        super(new KotlinJsr223DefaultScriptEngineFactory());
+        if (engineClassLoader != null) {
+            this.classLoader = engineClassLoader;
+        } else {
+            this.classLoader =
+                    new EngineClassLoader(
+                            getClass().getClassLoader(), ExtensionFactory.getAddOnLoader());
+        }
+    }
+
+    @Override
+    public ImageIcon getIcon() {
+        return ExtensionKotlin.KOTLIN_ICON;
+    }
+
+    @Override
+    public String getSyntaxStyle() {
+        return SyntaxConstants.SYNTAX_STYLE_NONE;
+    }
+
+    @Override
+    public boolean isRawEngine() {
+        return false;
+    }
+
+    @Override
+    public List<String> getExtensions() {
+        return Collections.singletonList("kts");
+    }
+
+    @Override
+    public ScriptEngine getEngine() {
+        return init(super.getEngine());
+    }
+
+    /**
+     * Initialises the script engine with custom class loader and function overrides.
+     *
+     * @param scriptEngine the script engine to initialise
+     * @return the script engine initialised
+     */
+    private ScriptEngine init(ScriptEngine scriptEngine) {
+        ClassLoader currentClassLoader = Thread.currentThread().getContextClassLoader();
+        Thread.currentThread().setContextClassLoader(classLoader);
+        try {
+            scriptEngine
+                    .getBindings(ScriptContext.ENGINE_SCOPE)
+                    .put("ZapScriptContext", scriptEngine.getContext());
+            // Note that this also forces the initialisation/usage of the custom class loader.
+            scriptEngine.eval(
+                    "fun print(msg: Any) { ZapScriptContext.writer.write(\"$msg\") }"
+                            + "fun println(msg: Any) { print(\"$msg${System.lineSeparator()}\") }");
```

Comment:
Does it really need to be OS dependent? (It can be added to the string, instead of called each time.)

Reply:
Yeah I'll put this back how it was, given where it's writing to I don't suspect it needs to be OS specific.